src/ranged.cpp

```diff
@@ -509,7 +504,7 @@ void player::fire_gun( const tripoint &targ_arg, bool burst )
         } else if ( used_weapon->deactivate_charger_gun() ) {
             // Deactivated charger gun
         } else {
-            used_weapon->charges -= used_weapon->ammo_required();
+            used_weapon->ammo_consume( used_weapon->ammo_required() );
```

Comment:
> Quitting on qty > ammo_remaining() sounds like a bug risk. You should probably print a debugmsg in such a case

That prevents use in a conditional, for example the idiom `while(foo.ammo_consume(n)) { ... }`. The pathological case (negative qty) is now handled separately printing a `debugmsg`.

> This is the only place where you use ammo_consume and you drop the value it returns.
> Shouldn't a debugmsg be printed on failure or something?

Agreed, the caller of `ammo_consume` should check the return and decide based upon context what to do if it returns `false`. I've added such a check in fb4c915 although I'm intending to refactor the firing loop in #14483 in any case; for example we shouldn't be ejecting shell casings before we consume ammo. I don't however want to make changes in `player::fire_gun` as part of this PR if at all avoidable.

> What about weapons that consume 0 ammo?

Now handled correctly in 5b87b07

On a tangent how much interest would there be for adding a cross-platform macro for `MUST_CHECK` implemented via `__attribute__ ((warn_unused_result))` and equivalent constructs?